src/rdb/rdb_tx.c

```diff
@@ -417,20 +420,58 @@ rdb_tx_commit(struct rdb_tx *tx)
 	 */
 	if (!rdb_tx_is_critical(tx)) {
 		daos_size_t	scm_remaining = 0;
+		uint32_t	nchecks = 0;
 
+check_space:
 		rc = rdb_scm_left(tx->dt_db, &scm_remaining);
 		if (rc != 0) {
 			D_ERROR(DF_DB": failed to query free space\n",
 				DP_DB(tx->dt_db));
 			goto out_lock;
 		}
+		nchecks++;
 
 		if (scm_remaining < RDB_NOAPPEND_FREE_SPACE) {
-			D_DEBUG(DB_TRACE, DF_DB": nearly out of space, do not "
-			       "append! scm_left="DF_U64"\n", DP_DB(tx->dt_db),
-			       scm_remaining);
-			D_GOTO(out_lock, rc = -DER_NOSPACE);
+			uint64_t		idx = 0;
+			uint64_t		delta_nsec;
+
+			if (nchecks > 1) {
+				D_DEBUG(DB_TRACE, DF_DB": nearly out of space, do not append! "
+				       "scm_left="DF_U64"\n", DP_DB(tx->dt_db), scm_remaining);
+				D_GOTO(out_lock, rc = -DER_NOSPACE);
+			} else {
+				/* Throttle number of times compacting the log when low on space. */
+				delta_nsec = (d_hlc2nsec(d_hlc_get()) -
+					      d_hlc2nsec(tx->dt_db->d_nospc_ts));
```

Comment:
[Question] It looks like these timestamps are only used in the current process. Why do we use the HLC, not the monotonic clock?

Reply:
Right, I think the non-coarse version of MONOTONIC clock is more suitable for the process-local use case here.